File: samples/CameraAccess/CameraAccess/WebRTC/SignalingClient.swift
import Foundation
import WebRTC

enum SignalingMessage {
  case roomCreated(String)
  case roomRejoined(String)
  case roomJoined
  case peerJoined
  case peerLeft
  case offer(RTCSessionDescription)
  case answer(RTCSessionDescription)
  case candidate(RTCIceCandidate)
  case error(String)
}

/// WebSocket client for WebRTC signaling. Exchanges SDP offers/answers and ICE candidates
/// via a relay server. Follows the same URLSessionWebSocketTask pattern as GeminiLiveService.
class SignalingClient {
  var onMessageReceived: ((SignalingMessage) -> Void)?
  var onConnected: (() -> Void)?
  var onDisconnected: ((String?) -> Void)?

  private var webSocketTask: URLSessionWebSocketTask?
  private var urlSession: URLSession!
  private let delegate = WebSocketDelegate()
  private let sendQueue = DispatchQueue(label: "signaling.send")
  private var receiveTask: Task<Void, Never>?

  init() {
    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    self.urlSession = URLSession(
      configuration: config, delegate: delegate, delegateQueue: nil)
  }

  func connect(url: URL) {
    disconnect()

    delegate.onOpen = { [weak self] in
      NSLog("[Signaling] Connected to %@", url.absoluteString)
      self?.onConnected?()
    }
    delegate.onClose = { [weak self] reason in
      NSLog("[Signaling] Disconnected: %@", reason ?? "unknown")
      self?.onDisconnected?(reason)
    }

    webSocketTask = urlSession.webSocketTask(with: url)
    webSocketTask?.resume()
    startReceiving()
  }

  func createRoom() {
    sendJSON(["type": "create"])
  }

  func joinRoom(code: String) {
    sendJSON(["type": "join", "room": code])
  }

  func rejoinRoom(code: String) {
    sendJSON(["type": "rejoin", "room": code])
  }

  func send(sdp: RTCSessionDescription) {
    let type = sdp.type == .offer ? "offer" : "answer"
    sendJSON(["type": type, "sdp": sdp.sdp])
  }

  func send(candidate: RTCIceCandidate) {
    sendJSON([
      "type": "candidate",
      "candidate": candidate.sdp,
      "sdpMid": candidate.sdpMid ?? "",
      "sdpMLineIndex": candidate.sdpMLineIndex,
    ] as [String: Any])
  }

  func disconnect() {
    receiveTask?.cancel()
    receiveTask = nil
    webSocketTask?.cancel(with: .goingAway, reason: nil)
    webSocketTask = nil
  }

  // MARK: - Private

  private func sendJSON(_ dict: [String: Any]) {
    sendQueue.async { [weak self] in
      guard let self, let task = self.webSocketTask else { return }
      guard let data = try? JSONSerialization.data(withJSONObject: dict),
        let text = String(data: data, encoding: .utf8)
      else { return }
      task.send(.string(text)) { error in
        if let error {
          NSLog("[Signaling] Send error: %@", error.localizedDescription)
        }
      }
    }
  }

  private func startReceiving() {
    receiveTask = Task { [weak self] in
      while !Task.isCancelled {
        guard let self, let task = self.webSocketTask else { break }
        do {
          let message = try await task.receive()
          switch message {
          case .string(let text):
            self.handleMessage(text)
          case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
              self.handleMessage(text)
            }
          @unknown default:
            break
          }
        } catch {
          if !Task.isCancelled {
            NSLog("[Signaling] Receive error: %@", error.localizedDescription)
            self.onDisconnected?(error.localizedDescription)
          }
          break
        }
      }
    }
  }

  private func handleMessage(_ text: String) {
    guard let data = text.data(using: .utf8),
      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let type = json["type"] as? String
    else { return }

    switch type {
    case "room_created":
      if let room = json["room"] as? String {
        onMessageReceived?(.roomCreated(room))
      }

    case "room_joined":
      onMessageReceived?(.roomJoined)

    case "room_rejoined":
      if let room = json["room"] as? String {
        onMessageReceived?(.roomRejoined(room))
      }

    case "peer_joined":
      onMessageReceived?(.peerJoined)

    case "peer_left":
      onMessageReceived?(.peerLeft)

    case "offer":
      if let sdp = json["sdp"] as? String {
        let sessionDesc = RTCSessionDescription(type: .offer, sdp: sdp)
        onMessageReceived?(.offer(sessionDesc))
      }

    case "answer":
      if let sdp = json["sdp"] as? String {
        let sessionDesc = RTCSessionDescription(type: .answer, sdp: sdp)
        onMessageReceived?(.answer(sessionDesc))
      }

    case "candidate":
      if let candidate = json["candidate"] as? String,
        let sdpMid = json["sdpMid"] as? String,
        let sdpMLineIndex = json["sdpMLineIndex"] as? Int32
      {
        let iceCandidate = RTCIceCandidate(
          sdp: candidate, sdpMLineIndex: sdpMLineIndex, sdpMid: sdpMid)
        onMessageReceived?(.candidate(iceCandidate))
      }

    case "error":
      let msg = json["message"] as? String ?? "Unknown signaling error"
      onMessageReceived?(.error(msg))

    default:
      NSLog("[Signaling] Unknown message type: %@", type)
    }
  }
}

// MARK: - WebSocket Delegate

private class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate {
  var onOpen: (() -> Void)?
  var onClose: ((String?) -> Void)?

  func urlSession(
    _ session: URLSession, webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    onOpen?()
  }

  func urlSession(
    _ session: URLSession, webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?
  ) {
    let reasonStr = reason.flatMap { String(data: $0, encoding: .utf8) }
    onClose?(reasonStr)
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/webrtc/SignalingClient.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.webrtc

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import org.webrtc.IceCandidate
import org.webrtc.SessionDescription
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

sealed class SignalingMessage {
    data class RoomCreated(val room: String) : SignalingMessage()
    data class RoomRejoined(val room: String) : SignalingMessage()
    object RoomJoined : SignalingMessage()
    object PeerJoined : SignalingMessage()
    object PeerLeft : SignalingMessage()
    data class Offer(val sdp: SessionDescription) : SignalingMessage()
    data class Answer(val sdp: SessionDescription) : SignalingMessage()
    data class Candidate(val candidate: IceCandidate) : SignalingMessage()
    data class Error(val message: String) : SignalingMessage()
}

class SignalingClient {
    companion object {
        private const val TAG = "SignalingClient"
    }

    var onMessageReceived: ((SignalingMessage) -> Unit)? = null
    var onConnected: (() -> Unit)? = null
    var onDisconnected: ((String?) -> Unit)? = null

    private var webSocket: WebSocket? = null
    private val sendExecutor = Executors.newSingleThreadExecutor()
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS) // No read timeout for WebSocket
        .connectTimeout(30, TimeUnit.SECONDS)
        .build()

    fun connect(url: String) {
        disconnect()

        val request = Request.Builder()
            .url(url)
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "Connected to $url")
                onConnected?.invoke()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "Closing: $reason")
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "Closed: $reason")
                onDisconnected?.invoke(reason.ifEmpty { null })
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "Failure: ${t.message}")
                onDisconnected?.invoke(t.message)
            }
        })
    }

    fun createRoom() {
        sendJSON(JSONObject().put("type", "create"))
    }

    fun joinRoom(code: String) {
        sendJSON(JSONObject().put("type", "join").put("room", code))
    }

    fun rejoinRoom(code: String) {
        sendJSON(JSONObject().put("type", "rejoin").put("room", code))
    }

    fun sendSdp(sdp: SessionDescription) {
        val type = if (sdp.type == SessionDescription.Type.OFFER) "offer" else "answer"
        sendJSON(JSONObject().put("type", type).put("sdp", sdp.description))
    }

    fun sendCandidate(candidate: IceCandidate) {
        sendJSON(
            JSONObject()
                .put("type", "candidate")
                .put("candidate", candidate.sdp)
                .put("sdpMid", candidate.sdpMid ?: "")
                .put("sdpMLineIndex", candidate.sdpMLineIndex)
        )
    }

    fun disconnect() {
        webSocket?.close(1000, "Going away")
        webSocket = null
    }

    private fun sendJSON(json: JSONObject) {
        sendExecutor.execute {
            val ws = webSocket ?: return@execute
            ws.send(json.toString())
        }
    }

    private fun handleMessage(text: String) {
        try {
            val json = JSONObject(text)
            val type = json.optString("type", "")

            when (type) {
                "room_created" -> {
                    val room = json.optString("room", "")
                    if (room.isNotEmpty()) {
                        onMessageReceived?.invoke(SignalingMessage.RoomCreated(room))
                    }
                }
                "room_joined" -> {
                    onMessageReceived?.invoke(SignalingMessage.RoomJoined)
                }
                "room_rejoined" -> {
                    val room = json.optString("room", "")
                    if (room.isNotEmpty()) {
                        onMessageReceived?.invoke(SignalingMessage.RoomRejoined(room))
                    }
                }
                "peer_joined" -> {
                    onMessageReceived?.invoke(SignalingMessage.PeerJoined)
                }
                "peer_left" -> {
                    onMessageReceived?.invoke(SignalingMessage.PeerLeft)
                }
                "offer" -> {
                    val sdp = json.optString("sdp", "")
                    if (sdp.isNotEmpty()) {
                        onMessageReceived?.invoke(
                            SignalingMessage.Offer(
                                SessionDescription(SessionDescription.Type.OFFER, sdp)
                            )
                        )
                    }
                }
                "answer" -> {
                    val sdp = json.optString("sdp", "")
                    if (sdp.isNotEmpty()) {
                        onMessageReceived?.invoke(
                            SignalingMessage.Answer(
                                SessionDescription(SessionDescription.Type.ANSWER, sdp)
                            )
                        )
                    }
                }
                "candidate" -> {
                    val candidate = json.optString("candidate", "")
                    val sdpMid = json.optString("sdpMid", "")
                    val sdpMLineIndex = json.optInt("sdpMLineIndex", 0)
                    if (candidate.isNotEmpty()) {
                        onMessageReceived?.invoke(
                            SignalingMessage.Candidate(
                                IceCandidate(sdpMid, sdpMLineIndex, candidate)
                            )
                        )
                    }
                }
                "error" -> {
                    val msg = json.optString("message", "Unknown signaling error")
                    onMessageReceived?.invoke(SignalingMessage.Error(msg))
                }
                else -> {
                    Log.w(TAG, "Unknown message type: $type")
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse signaling message: ${e.message}")
        }
    }
}
